lib/matplotlib/_layoutgrid.py

```diff
@@ -0,0 +1,577 @@
+            (self.bottoms[rows[-1]].value()),
+            (self.lefts[cols[0]].value() +
+                self.margins['leftcb'][cols[0]].value() +
+                self.margins['left'][cols[0]].value()),
+            (self.tops[rows[0]].value()))
+        return bbox
+
+    def get_bottom_margin_bbox(self, rows=0, cols=0):
+        """
+        Return the left margin bounding box of the subplot specs
+        given by rows and cols.  rows and cols can be spans.
+        """
+        rows = np.atleast_1d(rows)
+        cols = np.atleast_1d(cols)
+
+        bbox = Bbox.from_extents(
+            (self.lefts[cols[0]].value()),
+            (self.bottoms[rows[-1]].value() +
+             self.margins['bottomcb'][rows[-1]].value()),
+            (self.rights[cols[-1]].value()),
+            (self.bottoms[rows[-1]].value() +
+                self.margins['bottom'][rows[-1]].value() +
+             self.margins['bottomcb'][rows[-1]].value()
+             ))
+        return bbox
+
+    def get_right_margin_bbox(self, rows=0, cols=0):
+        """
+        Return the left margin bounding box of the subplot specs
+        given by rows and cols.  rows and cols can be spans.
+        """
+        rows = np.atleast_1d(rows)
+        cols = np.atleast_1d(cols)
+
+        bbox = Bbox.from_extents(
+            (self.rights[cols[-1]].value() -
+                self.margins['right'][cols[-1]].value() -
+                self.margins['rightcb'][cols[-1]].value()),
+            (self.bottoms[rows[-1]].value()),
+            (self.rights[cols[-1]].value() -
+                self.margins['rightcb'][cols[-1]].value()),
+            (self.tops[rows[0]].value()))
+        return bbox
+
+    def get_top_margin_bbox(self, rows=0, cols=0):
+        """
+        Return the left margin bounding box of the subplot specs
+        given by rows and cols.  rows and cols can be spans.
+        """
+        rows = np.atleast_1d(rows)
+        cols = np.atleast_1d(cols)
+
+        bbox = Bbox.from_extents(
+            (self.lefts[cols[0]].value()),
+            (self.tops[rows[0]].value() -
+                self.margins['topcb'][rows[0]].value()),
+            (self.rights[cols[-1]].value()),
+            (self.tops[rows[0]].value() -
+                self.margins['topcb'][rows[0]].value() -
+                self.margins['top'][rows[0]].value()))
+        return bbox
+
+    def update_variables(self):
+        """
+        Update the variables for the solver attached to this layoutgrid.
+        """
+        self.solver.updateVariables()
+
+_layoutboxobjnum = itertools.count()
+
+
+def seq_id():
+    """Generate a short sequential id for layoutbox objects."""
+    return '%06d' % next(_layoutboxobjnum)
+
+
+def nonetree(lb):
+    if lb.parent is None:
+        nonechildren(lb)
+        del(lb)
+    else:
+        nonetree(lb.parent)
+
+
+def nonechildren(lb):
+    if lb is None:
+        return
+    for child in lb.children.flat:
+        nonechildren(child)
+        del(child)
```

Comment:
I'm still not sure what effect you're aiming for here; `child` would have lost a reference after going out of scope, which is basically every iteration of the `for`-loop and the end of the function, which is exactly what this `del` covers. It does not affect `lb.children`.

Reply:
I'm trying to clear all the grid boxes on the tree so they aren't orphaned when we do clf.  Basically just stops memory leaks.  Is there a better way to do this?

Reply:
Right now, this whole function effectively does nothing. It recursively calls itself on all children, and for each child deletes its _local variable_ referencing it. Nothing is actually done to the children as they're still referenced in `lb.children`.

`nonetree` similarly also does nothing; it walks up the parents to the topmost one (while deleting its _local_ reference), and calls `nonechildren` on the topmost parent. But whatever called `nonetree` still has its reference to `lb`.

Reply:
OK< these were only used in one spot (`Figure.clf`) and I really don't think it was necessary there.  I just delete the parent layout grid, and then re-initialize `Figure._layoutgrid`.   I assume gc will take care of the rest, or deleting the children gridspecs.

Reply:
I'll squash the commit later, but the latest commit has the changes....